virtualmachine/vsphere/vm.go

```diff
@@ -856,84 +856,52 @@ func (vm *VM) Destroy() (err error) {
 		return nil
 	}
 
-	state, err := getState(vm)
+	vmMo, err := findVM(vm, getVMSearchFilter(vm.Name))
 	if err != nil {
 		return err
 	}
+	vmo := object.NewVirtualMachine(vm.client.Client, vmMo.Reference())
 
-	// Can't destroy a suspended VM, power it on and update the state
-	if state == "standby" {
-		err = start(vm)
-		if err != nil {
-			return err
-		}
+	powerState, err := getPowerState(vm)
+	if err != nil {
+		return err
 	}
 
-	if state != "notRunning" {
-		// Only possible states are running, shuttingDown, resetting or notRunning
-		timer := time.NewTimer(time.Second * 90)
-		wg := sync.WaitGroup{}
-		wg.Add(1)
-
-		go func() {
-			defer timer.Stop()
-			defer wg.Done()
-
-			powerState := "unInitialized"
-		Outerloop:
-			for {
-				state, e := getState(vm)
-				if e != nil {
-					err = e
-					break
-				}
-
-				if state == "notRunning" {
-					powerState, e = getPowerState(vm)
-					if e != nil {
-						err = e
-						break
-					}
-
-					if powerState == "poweredOff" {
-						break
-					}
-				}
-
-				if state == "running" {
-					e = halt(vm)
-					if e != nil {
-						err = e
-						break
-					}
-				}
-
-				select {
-				case <-timer.C:
-					err = fmt.Errorf("timed out waiting for VM to power off."+
-						"Current GuestState: %s and PowerState: %s", state, powerState)
-					break Outerloop
-				default:
-					// No action
-				}
-				time.Sleep(time.Second)
+	for powerState != "poweredOff" {
+		// Only possible states are poweredOff, poweredOn, suspended
+		if !isTaskInProgress(vm, vmMo) {
```

Comment:
Will it be better if we move this logic of waiting for tasks to finish to <isTaskInProgress>?

Reply:
done